File: app/src/main/java/com/liweiyap/narradir/util/fonts/FontCache.java
package com.liweiyap.narradir.util.fonts;

import android.content.Context;
import android.graphics.Typeface;

import androidx.annotation.NonNull;
import androidx.collection.ArrayMap;

import org.jetbrains.annotations.Nullable;

/**
 * Caches the custom Typefaces in the assets directory to:
 *  - avoid memory leaks on older handsets (https://issuetracker.google.com/issues/36919609)
 *  - increase execution speed at run time, since it's not super fast to read from assets all the time.
 */
public final class FontCache
{
    private FontCache(){}

    private static final ArrayMap<String, Typeface> fontCache = new ArrayMap<>();

    public static @Nullable Typeface get(final String assetFontPath, final @NonNull Context context)
    {
        if (assetFontPath == null)
        {
            return null;
        }

        Typeface typeface = fontCache.get(assetFontPath);
        
        if (typeface == null)
        {
            try
            {
                typeface = Typeface.createFromAsset(context.getAssets(), assetFontPath);
            }
            catch (Exception e)
            {
                return null;
            }
            
            fontCache.put(assetFontPath, typeface);
        }
        
        return typeface;
    }
}

File: app/src/main/java/com/liweiyap/narradir/util/fonts/FontCache.kt
package com.liweiyap.narradir.util.fonts

import android.content.Context
import android.graphics.Typeface

import androidx.collection.ArrayMap

import kotlin.Exception

/**
 * Caches the custom Typefaces in the assets directory to:
 * - avoid memory leaks on older handsets (https://issuetracker.google.com/issues/36919609)
 * - increase execution speed at run time, since it's not super fast to read from assets all the time.
 */
object FontCache {
    private val sFontCache = ArrayMap<String, Typeface>()

    @JvmStatic
    fun get(assetFontPath: String?, context: Context): Typeface? {
        if (assetFontPath == null) {
            return null
        }

        synchronized(sFontCache) {
            var typeface: Typeface? = sFontCache[assetFontPath]

            if (typeface == null) {
                try {
                    typeface = Typeface.createFromAsset(context.assets, assetFontPath)
                } catch (e: Exception) {
                    return null
                }

                sFontCache[assetFontPath] = typeface
            }

            return typeface
        }
    }
}